### PR TITLE
Merge of auto-config and auto-encrypt code

### PR DESCRIPTION
**Pull request by mkeeler**
auto-encrypt is now handled as a special case of auto-config.

This also is moving all the cert-monitor code into the auto-config package which is the bulk of the code changes. This allowed removing the whole Fallback callback cyclical dep from the BaseDeps type.

This PR was a net negative change by a few hundred lines but then I discovered some of the auto-encrypt functionality and some other auto-config functions were not as well covered by tests as I would like so I implemented those too and pushed this into a net addition of ~100 LOC.

Tasks:

- [x] Add `auto-encrypt` support to the `auto-config` package.
- [x] Update the `agent` to use `auto-config` for `auto-encrypt`
- [x] Remove the `consul` packages `auto-encrypt` code.
- [x] Migrate `cert-monitor` tests to equivalent tests in the `auto-config` package
- [x] Migrate `auto-encrypt` tests from the `consul` package into the `auto-config` package
- [x] Remove the `cert-monitor` package.
- [x] Move `router` creation out of the `Client` and `Server` types and into the `BaseDeps` creation. Depends on #8511. Then the common router can be passed into the `AutoConfig` type. This was merged in #8559

### CODE REVIEW

File: proto/translate.go [hunk 1]
```diff
@@ -12,14 +12,19 @@ var (
 	timePtrType = reflect.TypeOf((*time.Time)(nil))
 	timeType    = timePtrType.Elem()
 	mapStrInf   = reflect.TypeOf((map[string]interface{})(nil))
+
+	epoch1970 = time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)
```
Comment (mkeeler) [inline proto/translate.go:16]: The epoch1970 stuff is dealing with the fact that the protobuf zero ts is Jan 1 1970 where Go's time.Time has a zero value of Jan 1, 0. So we consider anything prior to jan 1 1970 to mean the same thing. In practice this wont be a problem.

File: agent/agent.go [hunk 2]
```diff
@@ -373,6 +370,11 @@ func New(bd BaseDeps) (*Agent, error) {
 	// pass the agent itself so its safe to move here.
 	a.registerCache()
 
+	// TODO: move to newBaseDeps
```
Comment (mkeeler) [inline agent/agent.go:373]: Maybe I could have moved this to BaseDeps? I moved it here before rebasing onto master which had the BaseDeps added.

Comment (dnephin) [reply agent/agent.go:373]: I started that work in #8554. It doesn't do the move just yet, but I think it does all the setup to make that possible.

Previously it happened after ` AutoConfig.InitialConfiguration`, but I guess this PR flips the order? I'll read over this now to get a better idea

File: agent/auto-config/config.go [hunk 3]
```diff
@@ -49,14 +78,78 @@ type Config struct {
...
+	FallbackRetry time.Duration
+
+	// Tokens is the shared token store. It is used to retrieve the current
+	// agent token as well as getting notifications when that token is updated.
+	// This field is required.
+	Tokens TokenStore
+}
+
+// WithLogger will cause the created AutoConfig type to use the provided logger
+func (c *Config) WithLogger(logger hclog.Logger) *Config {
```
Comment (dnephin) [inline agent/auto-config/config.go:108]: I think these `WithX` builder functions are no longer used since we populate the `Config struct directly. Do you think they will be needed or could be remove them? Maybe this came from resolving a merge conflcit?

Comment (mkeeler) [reply agent/auto-config/config.go:108]: I think we can remove them. They aren't used any more so we should get rid of the dead code.

File: agent/agent.go (hunk 2 above)
Comment (mkeeler) [reply agent/agent.go:373]: Before, AutoConfig never needed to know about the ACL token so loading after was fine.

Now AutoConfig might actually be performing auto-encrypt in which case it needs the acl token to authorize the request so it has to be loaded earlier. I think this temporary move is probably necessary until #8554 is finished up but it should be a pretty minor merge conflict to resolve.

File: agent/auto-config/config.go (hunk 3 above)
Comment (mkeeler) [reply agent/auto-config/config.go]: They have been removed.

File: agent/auto-config/run.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+			return fmt.Errorf("error while retrieving new agent certificates via auto-config: %w", err)
+		}
+
+		return ac.recordInitialResponse(resp)
+	case ac.config.AutoEncryptTLS:
+		reply, err := ac.autoEncryptInitialCerts(ctx)
+		if err != nil {
+			return fmt.Errorf("error while retrieving new agent certificate via auto-encrypt: %w", err)
+		}
+		return ac.setInitialTLSCertificates(reply)
```
Comment (i0rek) [inline agent/auto-config/run.go:101]: `auto_encrypt` doesn't store the initial response like `auto_config`, but I think it would be nice to have that. This is something that we could leave for another PR.

**Review by i0rek — APPROVED** [review]
LGTM!

File: agent/auto-config/run.go (hunk 4 above)
Comment (mkeeler) [reply agent/auto-config/run.go:101]: Yeah this already was kind of large. I was trying for this to be as pure of a refactor as I could make it without any behavioral modifications. If we did want to allow auto-encrypt to store to the filesystem it would be pretty easy in the future.

**Comment by mkeeler** [comment]
Backport merge conflicts will be resolved in #8589
